File: Dynamic Programming/279 Perfect Squares.cpp

//1.Dynamic Programming: 440ms
// Bottom-up
class Solution 
{
public:
    int numSquares(int n) 
    {
        vector<int> cntPerfectSquares(n + 1, INT_MAX);
        cntPerfectSquares[0] = 0;
        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j*j <= i; j++) // 这个会慢，因为有重复比如 13-9 = 4 又会做13-9 = 4, 做了两回的operation
            {
                cntPerfectSquares[i] = 
                    min(cntPerfectSquares[i], cntPerfectSquares[i - j*j] + 1);
            }
        }
        
        return cntPerfectSquares.back();
    }
};

// Bottom-up
class Solution {
public:
    int numSquares(int n) {
        vector<int>dp(n+1);
        iota(dp.begin(), dp.end(), 0);
        for(int i = 4; i <=n; ++i){
            for(int j = 2; j*j <= i; ++j){
                dp[i] = min(dp[i], dp[i-j*j]+1);
            }
        } 
        return dp[n];
    }
};

//Topdown 
class Solution {
public:
    int numSquares(int n) {
        vector<int>dp(n+1);
        iota(dp.begin(), dp.end(), 0);
        auto helper = [&](this auto&& helper, int i){
            if(i == 1) {
                return 1;
            }
            if (i == 0) {
                return 0;
            }
            if(dp[i]!= i || i <= 3)
                return dp[i];
            for(int j = 2; j*j <= i; ++j){
                dp[i] = min(dp[i], helper(i-j*j) + 1);
            }
            return dp[i];
        };
        return helper(n);
    }
};


// Bottom-up
class Solution { //更快的解 比起上面的 这个是 O(sqrt(n))*O(n/2)
public:
    int numSquares(int n) {
        vector<int>dp(n,-1);  
        for(int i = 1; i<=sqrt(n); i++){ 
            dp[i*i-1] = 1;
            for(int j = i*i+1; j<=n;j++){
                if (dp[j-1] == -1) dp[j-1] = j;
                dp[j-1]= min(dp[j-1],dp[j-i*i-1]+1);
            }
        }
        return dp[n-1];
    }
};


//Top-Down
class Solution {
public:
    int numSquares(int n) {
        vector<int>dp(n+1,-1);
        return topDown(dp, n);
    }
    
    int topDown(vector<int>&dp, int i){
        if(dp[i] >= 0)
            return dp[i];
        int cur = i;
        for(int j = 1; j*j<=i; ++j){
            cur = min(cur, topDown(dp, i-j*j)+1);
        }
        return dp[i] = cur;
    }
};






//2.Static Dynamic Programming: 12ms

class Solution { 
public:
    int numSquares(int n) {
        static vector<int>dp = {1};  //用了static dynamic programming
        if(dp.size()<n){
            int size = dp.size();
            vector<int>temp(n-dp.size(),-1);
            dp.insert(dp.end(),temp.begin(),temp.end());
            for(int i = 1; i<=sqrt(n); i++){ 
                dp[i*i-1] = 1;
                for(int j = max(i*i+1,size); j<=n;j++){
                    if (dp[j-1] == -1) dp[j-1] = j;
                    dp[j-1]= min(dp[j-1],dp[j-i*i-1]+1);
                }
            }
        }
        return dp[n-1];
    }
};







// BFS

class Solution 
{
public:
    int numSquares(int n)   
    {
        if (n <= 0)
        {
            return 0;
        }
        
        vector<int> perfectSquares;
        vector<int> cntPerfectSquares(n);
        queue<int> searchQ;

        for (int i = 1; i*i <= n; i++)
        {
            perfectSquares.push_back(i*i);
            searchQ.push(i*i);
            cntPerfectSquares[i*i - 1] = 1;
        }
        
        if (perfectSquares.back() == n)
        {
            return 1;
        }
        
        
        int currCntPerfectSquares = 1;
        while (!searchQ.empty())
        {
            currCntPerfectSquares++;
            
            int searchQSize = searchQ.size();
            for (int i = 0; i < searchQSize; i++)
            {
                int tmp = searchQ.front();

                for (auto& j : perfectSquares)
                {
                    if (tmp + j == n)
                    {
                        // We have reached node n.
                        return currCntPerfectSquares;
                    }
                    else if ((tmp + j < n) && (cntPerfectSquares[tmp + j - 1] == 0))
                    {
                        cntPerfectSquares[tmp + j - 1] = currCntPerfectSquares;
                        searchQ.push(tmp + j);
                    }
                    else if (tmp + j > n)
                    {
                        break;
                    }
                }
                
                searchQ.pop();
            }
        }
        
        return 0;
    }
};

File: Dynamic Programming/279. Perfect Squares.py
"""
279. Perfect Squares
"""

#bottom-up
class Solution(object):
    _num = [0]
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        num = self._num
        while len(num) <= n:
            num += min(num[-i*i] for i in range(1, int(len(num)**0.5+1))) + 1,
        return num[n]


#bottom-up
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [n+1 for _ in range(n)]
        i = 1
        while i*i <= n:
            dp[i*i-1] = 1
            j = 1 
            for j in range(i*i+1, n+1):
                dp[j-1]= min(dp[j-1], dp[j-i*i-1]+1)
            i+=1
        return dp[-1]


#top-down
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [-1 for i in range(n+1)]
        dp[0] = 0
        def topDown(i):
            if i < 0: 
                return n
            if dp[i] >= 0:
                return dp[i]
            j = 1
            cur = i+1
            while j*j <= i:
                cur = min(cur, topDown(i-j*j)+1)
                j+=1
            dp[i] = cur
            return dp[i]
        return topDown(n)


#Static DP
class Solution:
    def __init__(self,):
        self.dp = [0,1]
        self.start = 1
    
    def numSquares(self, n: int) -> int:
        if n <= self.start:
            return self.dp[n]
        self.dp += [0 for _ in range(n - self.start)]
        for i in range(self.start + 1, n+1):
            cur = i
            j = 1
            while j*j <= n and i - j*j >= 0:
                cur = min(cur, 1 + self.dp[i - j*j])
                j += 1
            self.dp[i] = cur
        self.start = n
        return self.dp[n]

# BFS

class Solution:
    def numSquares(self, n: int) -> int:
        q = collections.deque()
        square = []
        i = 0
        while i*i <= n:
            square += i*i,
            q.append(i*i)
            i+=1
        if square[-1] == n:
            return 1
        dp = {}
        dp[0] = 1
        step = 1
        while q:
            size = len(q)
            step += 1
            for _ in range(size):
                cur = q.popleft()
                for i in square:
                    if cur + i == n:
                        return step
                    if i + cur > n:
                        break
                    if i + cur < n and i + cur not in dp:
                        dp[i+cur] = step
                        q.append(i+cur)

# Static DP
class Solution:
    _dp = [0]
    def numSquares(self, n: int) -> int:
        dp = self._dp
        while len(dp) <= n:
            dp += min(dp[-i*i] for i in range(1, int(len(dp)**0.5+1))) + 1,
        return dp[n]